File: 03-Stack/yujin0419/C-2504.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
	string str;
	stack<char> bracket;

	cin >> str;

	int answer = 0, temp = 1;
	for ( int i = 0; i <str.length(); i++) {
		if (str[i] == '(') {
			temp *= 2;
			bracket.push('(');
		}
		else if (str[i] == '[') {
			temp *= 3;
			bracket.push('[');
		}
		else if (str[i] == ')') {
			if (bracket.empty() || bracket.top() != '(') {
				answer = 0;
				break;
			}
			if (str[i - 1] == '(') {
				answer += temp;
				temp /= 2;
				bracket.pop();
			}
			else {
				temp /= 2;
				bracket.pop();
			}
		}
		else if (str[i] == ']') {
			if (bracket.empty() || bracket.top() != '[') {
				answer = 0;
				break;
			}
			if (str[i - 1] == '[') {
				answer += temp;
				temp /= 3;
				bracket.pop();
			}
			else {
				temp /= 3;
				bracket.pop();
			}
		}
	}

	if (!bracket.empty()) answer = 0;
	cout << answer << "\n";

	return 0;
}

File: 03-Stack/tlswldus03/C-2504.py
import sys
input = sys.stdin.readline
from collections import deque


stack = deque()
brackets = list(input().rstrip())
temp = 1
result = 0
before = ""

for bracket in brackets:
  if(bracket == "("):
    temp *= 2
    stack.append("(")
  elif(bracket == "["):
    temp *= 3
    stack.append("[")
  elif(bracket == ")"):
    if(stack and stack[-1] == "("):
      stack.pop()
      if(before == "("):
        result += temp
      temp = temp // 2
    else:
      stack.append(bracket)
  elif(bracket == "]"):
    if(stack and stack[-1] == "["):
      stack.pop()
      if(before == "["):
        result += temp
      temp = temp //3
    else:
      stack.append(bracket)
  before = bracket
      
      
if(stack):
  print("0")
else:
  print(result)
  